packages/core/src/sso/SamlConnector/index.ts

```diff
@@ -0,0 +1,174 @@
+import {
+  ConnectorError,
+  ConnectorErrorCodes,
+  type GetSession,
+  type SetSession,
+} from '@logto/connector-kit';
+import { assert, appendPath } from '@silverhand/essentials';
+import * as saml from 'samlify';
+import { z } from 'zod';
+
+import { EnvSet, getTenantEndpoint } from '#src/env-set/index.js';
+
+import {
+  type BaseSamlConfig,
+  type BaseSamlConnectorConfig,
+  attributeMappingPostProcessor,
+} from '../types/saml.js';
+
+import {
+  fetchSamlConfig,
+  getRawSamlConfig,
+  getUserInfoFromRawUserProfile,
+  samlAssertionHandler,
+} from './utils.js';
+
+class SamlConnector {
+  private readonly _acsUrl: string;
+  constructor(
+    private readonly config: BaseSamlConnectorConfig,
+    tenantId: string,
+    ssoConnectorId: string
+  ) {
+    this._acsUrl = appendPath(
+      getTenantEndpoint(tenantId, EnvSet.values),
+      `api/authn/saml/sso/${ssoConnectorId}`
+    ).toString();
+  }
+
+  get acsUrl() {
+    return this._acsUrl;
+  }
+
+  /* Fetch SAML config from the metadata XML file or metadata URL. Throws error if config is invalid. */
+  getSamlConfig = async (): Promise<BaseSamlConfig> => {
```

Comment:
prefer use `async getSamlConfig() {}` over arrow functions in class. arrow functions cannot use `.bind()`